federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+			if !portsEqualForLB(svc, service) {
+				needUpdate = true
+			}
+			if needUpdate {
+				_, err = client.Core().Services(svc.Namespace).UpdateStatus(svc)
+				if err == nil {
+					glog.V(5).Infof("service %s/%s successfully updated to cluster %s", svc.Namespace, svc.Name, clusterName)
+					return nil
+				} else {
+					glog.V(4).Infof("failed to update status %+v", err)
+				}
+			} else {
+				glog.V(5).Infof("service %s/%s is not updated to cluster %s as the spec are identical", svc.Namespace, svc.Name, clusterName)
+				return nil
+			}
+		}
+		if svc == nil || errors.IsNotFound(err) {
+			// Create service if it is not found
+			glog.Infof("Service '%s/%s' does not exist in cluster %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			service.ResourceVersion = ""
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				glog.V(5).Infof("service %s/%s successfully created to cluster %s", service.Namespace, service.Name, clusterName)
+				return nil
+			}
+			glog.V(4).Infof("failed to create status %+v", err)
+		}
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			// ignore and get LB info
+			return nil
+		}
+		glog.Warningf("Failed to update service '%s/%s' in cluster %s: %v",
+			service.Namespace, service.Name, clusterName, err )
+		time.Sleep(clientRetryInterval)
+	}
+	return err
+}
+
+// ListKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *serviceCache) allServices() []*cachedService {
+	s.rwlock.Lock()
+	defer s.rwlock.Unlock()
+	services := make([]*cachedService, 0, len(s.fedServiceMap))
+	for _, v := range s.fedServiceMap {
+		services = append(services, v)
+	}
+	return services
+}
+
+func (s *serviceCache) get(serviceName string) (*cachedService, bool) {
+	s.rwlock.Lock()
+	defer s.rwlock.Unlock()
+	service, ok := s.fedServiceMap[serviceName]
+	return service, ok
+}
+
+func (s *serviceCache) getOrCreate(serviceName string) *cachedService {
+	s.rwlock.Lock()
+	defer s.rwlock.Unlock()
+	service, ok := s.fedServiceMap[serviceName]
+	if !ok {
+		service = &cachedService {
+			endpointMap: make(map[string]int),
+			serviceStatusMap: make(map[string]v1.LoadBalancerStatus),
+		}
+		s.fedServiceMap[serviceName] = service
+	}
+	return service
+}
+
+func (s *serviceCache) set(serviceName string, service *cachedService) {
+	s.rwlock.Lock()
+	defer s.rwlock.Unlock()
+	s.fedServiceMap[serviceName] = service
+}
+
+func (s *serviceCache) delete(serviceName string) {
+	s.rwlock.Lock()
+	defer s.rwlock.Unlock()
+	delete(s.fedServiceMap, serviceName)
+}
+
+func (s *ServiceController) needsUpdate(oldService *v1.Service, newService *v1.Service) bool {
+	if !wantsDNSRecords(oldService) && !wantsDNSRecords(newService) {
+		return false
```

Comment:
this is validation function, so there is no write or delete function.
I assume this function should link with dns.go:ensureDNSRecords
one for checking, the other for operation.